File: src/gbaemu/gba/keypad.cpp
#include <cstdint>
#include <cstdio>

#include <gbaemu/gba/io.hpp>
#include <gbaemu/gba/keypad.hpp>

namespace gbaemu::gba::keypad {
    void init() {
        gbaemu::gba::io::set(io::KEYINPUT, 0x03ff);
        gbaemu::gba::io::set(io::KEYCNT, 0x0000);
    }

    void keyDown(key_t key) {
        uint16_t keyinput = gbaemu::gba::io::get(io::KEYINPUT);
        keyinput &= ~key;
        gbaemu::gba::io::set(io::KEYINPUT, keyinput);
    }

    void keyUp(key_t key) {
        uint16_t keyinput = gbaemu::gba::io::get(io::KEYINPUT);
        keyinput |= key;
        gbaemu::gba::io::set(io::KEYINPUT, keyinput);
    }
}

File: src/GBA/Keypad.cs
namespace gbaemu.GBA {
    public class Keypad {
        public GBA GBA {get; private set;}

        public enum Keys {
            A,
            B,
            Select,
            Start,
            Right,
            Left,
            Up,
            Down,
            R,
            L
        }

        public Keypad(GBA gba) {
            GBA = gba;
        }

        public void UpdatePressedKeys(bool[] pressedKeys) {
            ushort value = 0xffff;

            for(int i = 0; i < pressedKeys.Length; i++) {
                if(pressedKeys[i]) {
                    value &= (ushort)~(1 << i);
                }
            }

            IO.Register keyinput = GBA.Io.GetRegister(0x04000130);
            keyinput.Value = value;
        }

        public void Cycle() {
            IO.Register keyinput = GBA.Io.GetRegister(0x04000130);
            IO.Register keycnt = GBA.Io.GetRegister(0x04000132);
            
            if(BitUtils.BitTest16(keycnt.Value, 14)) {
                if(BitUtils.BitTest16(keycnt.Value, 15)) {
                    if((keycnt.Value & (~keyinput.Value) & 0x03ff) == (keycnt.Value & 0x03ff)) {
                        GBA.SetInterruptFlag(1 << 12);
                    }
                } else {
                    if((keycnt.Value & (~keyinput.Value) & 0x03ff) != 0) {
                        GBA.SetInterruptFlag(1 << 12);
                    }
                }
            }
        }
    }
}